File: day-7/day-7.py
#!/usr/bin/env python
################################################################################
# Copyright (C) 2018 Michael Wiseman                                           #
#                                                                              #
# This program is free software: you can redistribute it and/or modify it      #
# under the terms of the GNU Affero General Public License as published by the #
# Free Software Foundation, either version 3 of the License, or (at your       #
# option) any later version.                                                   #
#                                                                              #
# This program is distributed in the hope that it will be useful, but WITHOUT  #
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or        #
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License #
# for more details.                                                            #
#                                                                              #
# You should have received a copy of the GNU Affero General Public License     #
# along with this program.  If not, see <https://www.gnu.org/licenses/>.       #
################################################################################
import re
import string
from collections import defaultdict

# Function definitions

def part_one(input):
    s_input = sorted(input)
    letter_pairs_forward = defaultdict(list)
    letter_pairs_backward = defaultdict(list)
    letters = string.ascii_uppercase
    middle = ' must be finished before step '
    for line in s_input:
        aline = re.sub(middle, '=>', line)
        aline = re.sub('Step ', '', aline)
        aline = re.sub(' can begin.', '', aline)
        f,s = aline[0], aline[-1]
        if f not in letter_pairs_forward:
            letter_pairs_forward[f] = [s]
        else:
            letter_pairs_forward[f].append(s)
        if s not in letter_pairs_backward:
            letter_pairs_backward[s] = [f]
        else:
            letter_pairs_backward[s].append(f)
        letter_pairs_forward[f] = sorted(letter_pairs_forward[f])
        letter_pairs_backward[s] = sorted(letter_pairs_backward[s])
    for ltr in letters:
        if ltr not in letter_pairs_forward:
            letter_pairs_forward[ltr] =['']
        if ltr not in letter_pairs_backward:
            letter_pairs_backward[ltr] =['']
    for key in sorted(letter_pairs_backward.keys()):
        print(key, ':', letter_pairs_backward[key])

# def part_two(input):

input_file = open('day-7-input.txt', 'r').read().splitlines()

# Did both parts with pencil and paper mostly
part_one(input_file) # BGKDMJCNEQRSTUZWHYLPAFIVXO
# part_two(input_file) # 941

File: day-7/day-7.pl
#!/usr/bin/env perl
################################################################################
# Copyright (C) 2018 Michael Wiseman                                           #
#                                                                              #
# This program is free software: you can redistribute it and/or modify it      #
# under the terms of the GNU Affero General Public License as published by the #
# Free Software Foundation, either version 3 of the License, or (at your       #
# option) any later version.                                                   #
#                                                                              #
# This program is distributed in the hope that it will be useful, but WITHOUT  #
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or        #
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License #
# for more details.                                                            #
#                                                                              #
# You should have received a copy of the GNU Affero General Public License     #
# along with this program.  If not, see <https://www.gnu.org/licenses/>.       #
################################################################################
use strict;
use warnings;
use feature 'say';
use Data::Dumper;

# Function definitions

sub part_one
{
    my @input = @{$_[0]};
    my @s_input = sort @input;
    my %letter_pairs_forward;
    my %letter_pairs_backward;
    my @letters = 'A'..'Z';
    my $middle = ' must be finished before step ';

    for my $line (@s_input)
    {
        my $aline = $line =~ s/$middle/=>/r;
        $aline =~ s/Step //;
        $aline =~ s/ can begin\.//;
        my ($f, $s) = split /=>/, $aline;
        unless (exists $letter_pairs_forward{$f})
        {
            $letter_pairs_forward{$f} = [$s];
        }
        else
        {
            push @{$letter_pairs_forward{$f}}, $s;
        }
        unless (exists $letter_pairs_backward{$s})
        {
            $letter_pairs_backward{$s} = [$f];
        }
        else
        {
            push @{$letter_pairs_backward{$s}}, $f;
        }
    }
    for my $ltr (@letters)
    {
        if (defined $letter_pairs_backward{$ltr})
        {
            say $ltr, ': [', join(', ', @{$letter_pairs_backward{$ltr}}), ']';
        }
        else
        {
            say "$ltr: []";
        }

    }
}

# sub part_two
# {
# }

# Did both parts with pencil and paper mostly

my $filename = 'day-7-input.txt';
open(my $ifh, '<', $filename);
my @input_file = <$ifh>;
close $ifh;
chomp(@input_file);
part_one(\@input_file);
# part_two(\@input_file);
